cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AccessEnforcer.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright © 2016-2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.spi.authorization;
+
+import io.cdap.cdap.api.annotation.Beta;
+import io.cdap.cdap.proto.id.EntityId;
+import io.cdap.cdap.proto.security.Permission;
+import io.cdap.cdap.proto.security.Principal;
+import io.cdap.cdap.security.spi.AccessException;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Enforces authorization for a {@link Principal} to perform an {@link Permission} on an {@link EntityId}.
+ */
+@Beta
+public interface AccessEnforcer {
+
+  /**
+   * Enforces authorization for the specified {@link Principal} for the specified {@link Permission} on the specified
+   * {@link EntityId}.
+   *
+   * @param entity the {@link EntityId} on which authorization is to be enforced
+   * @param principal the {@link Principal} that performs the permission
+   * @param permission the {@link Permission} being performed
+   * @throws UnauthorizedException if the principal is not authorized to perform the specified permission on the entity
+   */
+  default void enforce(EntityId entity, Principal principal, Permission permission) throws AccessException {
+    enforce(entity, principal, Collections.singleton(permission));
+  }
+
+  /**
+   * Enforces authorization for the specified {@link Principal} for the specified {@link Permission permissions} on the
+   * specified {@link EntityId}.
+   *
+   * @param entity the {@link EntityId} on which authorization is to be enforced
+   * @param principal the {@link Principal} that performs the permissions
+   * @param permissions the {@link Permission permissions} being performed
+   * @throws UnauthorizedException if the principal is not authorized to perform the specified permissions on the entity
+   */
+  void enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions) throws AccessException;
```

Comment:
nit: Likely not very feasible short-term, but maybe we can consider refactoring the enforcement SPI to return a response instead of throwing exceptions for unauthorized checks since generally, exceptions shouldn't be used for common situations.

Reply:
We can consider it.